rofl-scheduler/src/proxy/domain.rs

```diff
@@ -0,0 +1,347 @@
+                        } else if v.retry_at.is_none_or(|retry_time| now >= retry_time) {
+                            ready_retries.push(v.clone());
+                            false
+                        } else {
+                            true
+                        }
+                    });
+
+                    for verification in ready_retries {
+                        let _ = work_tx.send(verification).await;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Run a worker task that processes domain verifications.
+    async fn run_worker(
+        work_rx: tokio_mpmc::Receiver<DomainVerification>,
+        retry_tx: mpsc::UnboundedSender<DomainVerification>,
+        notifier: Arc<dyn CustomDomainVerifierNotifier>,
+        logger: slog::Logger,
+    ) -> Result<()> {
+        let resolver = hickory_resolver::Resolver::builder_tokio()?.build();
+
+        while let Some(verification) = work_rx.recv().await? {
+            // Skip cancelled verifications.
+            if verification.is_cancelled() {
+                continue;
+            }
+
+            slog::info!(logger, "processing domain verification";
+                "domain" => &verification.domain,
+                "instance_id" => ?verification.instance_id,
+            );
+
+            // Perform a single verification attempt.
+            match Self::verify_domain(&resolver, &verification).await {
+                Ok(_) => {
+                    slog::info!(logger, "domain verification successful";
+                        "domain" => &verification.domain,
+                        "instance_id" => ?verification.instance_id,
+                    );
+
+                    if !verification.is_cancelled() {
+                        notifier
+                            .verification_completed(verification.instance_id, &verification.domain)
+                            .await;
+                    }
+                }
+                Err(err) => {
+                    slog::warn!(logger, "domain verification error, scheduling retry";
+                        "domain" => &verification.domain,
+                        "instance_id" => ?verification.instance_id,
+                        "err" => ?err,
+                    );
+                    let _ = retry_tx.send(verification);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Verify a single domain by checking its TXT records.
+    ///
+    /// This method performs a DNS TXT record lookup for the given domain and
+    /// checks if it contains the expected verification token for the instance.
+    async fn verify_domain(
+        resolver: &hickory_resolver::TokioResolver,
+        verification: &DomainVerification,
+    ) -> Result<()> {
+        let expected_token = format!("oasis-rofl-verification={}", verification.token);
+
+        let txt_records = resolver.txt_lookup(&verification.domain).await?;
```

Comment:
Can we trust the response from this lookup?

Reply:
We can also enable DNSSEC validation.